### ISSUE TITLE
Evaluate Data Imbalance Impact

### ISSUE THREAD

**Issue by paul-tqh-nguyen** [paul-tqh-nguyen/reuters_topic_labelling#19]
Our dataset is extremely unbalanced. 

It's not clear how to evaluate the impact of it outside of trying several common techniques and seeing what methods provide improvement.

We weren't able to find a whole lot of information or recommendations online for how to oversample sanely for multi-label data. Here's what we could find:
- Random Oversampling: This will work. We just have to make sure that we oversample just the minority class, which is obvious for binary classification but less obvious for multi-label classification.
- SMOTE: This would require us to generate synthetic samples. It's not exactly clear how we would do this for NLP data since we're learning the representations as well. 
    - We'd additionally still have to know what the minority class would be. 
    - Raw text data isn't continuous, so it's not clear that SMOTE can work. 
    - It's not clear how to generate new samples on non-fixed length sequences of vectors. 
    - Some have suggested converting the sequence into a fixed-length vector and then generating synthetic samples for those. This isn't really helpful for us because the sampling will only benefit the attention and prediction layer of our neural network. 
- Random Undersampling: Same applies here as does to Random Oversampling. 

There doesn't seem to be a standard way of handling this both for NLP in general, NLP via Deep Learning, or for multi-label data.

Perhaps it's not a problem in Deep Learning for some reason I've yet to figure out?

People have worked around this problem in other ways: 
- data augmentation via synonyms from a dictionary
- data augmentation by replacing embeddings with nearest neighbor embedding
- random NMT back and forth translations
- using decision trees and boosting as they are supposedly robust against data imbalance
- Using an algorithm that weighs under-represented or poorly classified classes, e.g. Adaboost
- Using cost-sensitive learning (this might actually be a better solution and not just a workaround)

Here's a proposal that might work:
- We effectively want the distribution of all label occurrences (labels can come from the same sample) to be a uniform distribution. 
- We can use the Chi Squared Test as a metric to see how far off we are from a uniform distribution. 
  - The Kolmogorov–Smirnov test is another option worth investigating https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test
  - Another idea was to take the mean of the occurrences for each label, make a cost that is the SSD from the mean and using that as a cost. 
- Let's oversample or undersample until the Chi Squared statistic is minimized.

How to implement this is sort of an open problem at the moment. 

We had two options:
1. Random / Greedy search. Iteratively add samples that decrease the Chi Squared statistic. 
2. Learn the occurrences of each label to minimize the Chi Squared metric. Each sample has a set of multi-hot output labels. The final distribution is the sum of these output labels. The Chi Squared metric is differentiable. The Chi Squared of the final distribution can be optimized to be low. We can learn coefficients that we multiply by each multi-hot output label that represent "how many times should this example occur to achieve a low Chi Squared metric?". Perhaps this will additionally tell us if we should undersample or not. We would need to add regularization to make sure all the coefficients don't just become zero.

Let's investigate this and see what results we get.

**Comment by paul-tqh-nguyen** [comment]
Here's some progress we've already made for a greedy / random search approach.
Summary:
```
# Linear Greedy
Current number of examples 19019
Current Chi-Squared Statistic: 13203.980080222995
Current Chi-Squared Statistic: 12906.325761366288
Execution took 33.281893730163574 seconds.

# Quadratic
Current number of examples 20169
Current Chi-Squared Statistic: 10123.380788919001
Execution took 35.78880071640015 seconds.


example_histogram = histogram(training_examples)
occurence_histogram = histogram(example_histogram.values())
p1(sorted(occurence_histogram.items(), key = lambda x: x[0]))
```
Preamble:
```
from main import *
from  models import *

classifier = EEAPClassifier(NUMBER_OF_EPOCHS, BATCH_SIZE, TRAIN_PORTION, VALIDATION_PORTION, TESTING_PORTION, MAX_VOCAB_SIZE, PRE_TRAINED_EMBEDDING_SPECIFICATION, ENCODING_HIDDEN_SIZE, NUMBER_OF_ENCODING_LAYERS, ATTENTION_INTERMEDIATE_SIZE, NUMBER_OF_ATTENTION_HEADS, DROPOUT_PROBABILITY, OUTPUT_DIR)

example_numericalized_labels = lambda example: torch.tensor([bool(getattr(example, topic)) for topic in classifier.topics], dtype=int)

all_examples = [e for e in classifier.all_data]
all_data_label_histogram = torch.tensor([example_numericalized_labels(example) for example in all_examples], dtype=int).sum(dim=0)

training_examples = [e for e in classifier.training_data]
training_data_label_histogram = torch.tensor([example_numericalized_labels(example) for example in training_examples], dtype=int).sum(dim=0)

validation_examples = [e for e in classifier.validation_data]
validation_data_label_histogram = torch.tensor([example_numericalized_labels(example) for example in validation_examples], dtype=int).sum(dim=0)

testing_examples = [e for e in classifier.testing_data]
testing_data_label_histogram = torch.tensor([example_numericalized_labels(example) for example in testing_examples], dtype=int).sum(dim=0)

label_counts_too_small_to_reach_training_data = sorted([e.item() for e in all_data_label_histogram*((training_data_label_histogram == 0).int()) if e.item() != 0])
label_counts_too_small_to_reach_validation_data = sorted([e.item() for e in all_data_label_histogram*((validation_data_label_histogram == 0).int()) if e.item() != 0])
label_counts_too_small_to_reach_testing_data = sorted([e.item() for e in all_data_label_histogram*((testing_data_label_histogram == 0).int()) if e.item() != 0])
```
Adaptive:
```
from scipy.stats import chisquare

training_examples = [e for e in classifier.training_data]

import multiprocessing

#

with timer():
    number_of_new_examples_per_iteration = len(classifier.training_data) // 10*2
    previous_chi_squared_statistic = float('inf')
    training_data_label_histogram = torch.stack([example_numericalized_labels(example) for example in training_examples]).sum(dim=0)
    for i in range(2000):
        print('\n')
        print(f'Iteration {i}')
        print(f'Current number of examples {len(training_examples)}')
        def chi_squared_statistic_upon_addition(example):
            return chisquare( example_numericalized_labels(example) + training_data_label_histogram ).statistic
        current_chi_squared_statistic = chisquare(training_data_label_histogram).statistic
        print(f'Current Chi-Squared Statistic: {current_chi_squared_statistic}')
        scores = parallel_map(chi_squared_statistic_upon_addition, classifier.training_data)
        score_example_pairs = zip(scores, training_examples)
        sorted_score_example_pairs = sorted(score_example_pairs, key=lambda x: x[0])
        best_examples = eager_map(lambda pair: pair[1], sorted_score_example_pairs[:number_of_new_examples_per_iteration])
        updated_data_label_histogram = torch.stack([example_numericalized_labels(example) for example in training_examples+best_examples]).sum(dim=0)
        if current_chi_squared_statistic > chisquare(updated_data_label_histogram).statistic:
            previous_chi_squared_statistic = current_chi_squared_statistic
            training_examples = training_examples + best_examples
            training_data_label_histogram = updated_data_label_histogram
            number_of_new_examples_per_iteration = number_of_new_examples_per_iteration + number_of_new_examples_per_iteration//10
        else:
            number_of_new_examples_per_iteration //= 2
```
Linear Batch:
```
from scipy.stats import chisquare

training_examples = [e for e in classifier.training_data]

import multiprocessing

#

with timer():
    for i in range(20):
        print('\n')
        print(f'Current number of examples {len(training_examples)}')
        training_data_label_histogram = torch.stack([example_numericalized_labels(example) for example in training_examples]).sum(dim=0)
        def example_is_worth_oversampling(example):
            chi_squared_statistic_open_addition = chisquare( example_numericalized_labels(example) + training_data_label_histogram ).statistic
            return example if chi_squared_statistic_open_addition else None
        print(f'Current Chi-Squared Statistic: {chisquare(training_data_label_histogram).statistic}')
        oversampled_examples = filter(identity, parallel_map(example_is_worth_oversampling, classifier.training_data))
        training_examples = training_examples + list(oversampled_examples)
```
Linear Greedy:
```
from scipy.stats import chisquare

training_examples = [e for e in classifier.training_data]

import multiprocessing

#

with timer():
    training_data_label_histogram = torch.stack([example_numericalized_labels(example) for example in training_examples]).sum(dim=0)
    for i in range(20):
        print('\n')
        print(f'Current number of examples {len(training_examples)}')
        current_chi_squared_statistic = chisquare(training_data_label_histogram).statistic
        print(f'Current Chi-Squared Statistic: {chisquare(training_data_label_histogram).statistic}')
        def chi_squared_statistic_upon_addition(example):
            return chisquare( example_numericalized_labels(example) + training_data_label_histogram ).statistic
        for example in classifier.training_data:
            chi_squared_statistic_upon_addition_of_example = chi_squared_statistic_upon_addition(example)
            if chi_squared_statistic_upon_addition_of_example < current_chi_squared_statistic:
                training_examples.append(example)
                current_chi_squared_statistic = chi_squared_statistic_upon_addition_of_example
                training_data_label_histogram += example_numericalized_labels(example)
    print(f'Current Chi-Squared Statistic: {chisquare(training_data_label_histogram).statistic}')
```
Quadratic:
```
from scipy.stats import chisquare

training_examples = [e for e in classifier.training_data]

import multiprocessing

#

with timer():
    for i in range(20):
        print('\n')
        print(f'Current number of examples {len(training_examples)}')
        training_data_label_histogram = torch.stack([example_numericalized_labels(example) for example in training_examples]).sum(dim=0)
        def chi_squared_statistic_upon_addition(example):
            return chisquare( example_numericalized_labels(example) + training_data_label_histogram ).statistic
        print(f'Current Chi-Squared Statistic: {chisquare(training_data_label_histogram).statistic}')
        scores = parallel_map(chi_squared_statistic_upon_addition, classifier.training_data)
        score_example_pairs = zip(scores, training_examples)
        sorted_score_example_pairs = sorted(score_example_pairs, key=lambda x: x[0])
        best_examples = eager_map(lambda pair: pair[1], sorted_score_example_pairs[:len(classifier.training_data)//10])
        training_examples = training_examples + best_examples
```

**Comment by paul-tqh-nguyen** [comment]
Here's an implementation that finds the number of times each sample should be repeated to get a Chi Squared p value of 1. 

We'd still have to ensure that two rows with identical labels had the same number of occurrences, but that's something we can figure out later with ease. 

It only takes 36.985456228256226 seconds.
```
from scipy.stats import chisquare

import torch
import torch.nn as nn
import torch.optim as optim

def chi_squared_loss(histogram: torch.Tensor) -> torch.Tensor:
    expected_value = torch.mean(histogram)
    loss = torch.sum((histogram-expected_value)**2)/expected_value
    return loss

def normalize_occurences(unnormalized_occurences: torch.Tensor) -> torch.Tensor:
    positive_occurences = torch.exp(unnormalized_occurences)
    greater_than_one_occurences = positive_occurences + 1
    return greater_than_one_occurences

single_occurence_histogram_matrix = torch.stack([example_numericalized_labels(example) for example in classifier.training_data]).to(DEVICE)
occurences = nn.Parameter( torch.log(torch.tensor(len(classifier.training_data), dtype=float)) * torch.rand((len(classifier.training_data),1)).to(DEVICE) )
loss_function = chi_squared_loss
optimizer = optim.Adam([occurences])

loss = float('inf')
chi_squared_p_value = 0
with timer():
    while 1-chi_squared_p_value > 1e-4:
        optimizer.zero_grad()
        normalized_occurences = normalize_occurences(occurences)
        total_histogram = (single_occurence_histogram_matrix * normalized_occurences).sum(dim=0)
        loss = chi_squared_loss(total_histogram)
        loss.backward()
        optimizer.step()
        learned_final_histogram = (single_occurence_histogram_matrix * normalized_occurences).sum(dim=0).to('cpu').detach()
        chi_squared_p_value = chisquare(learned_final_histogram).pvalue
```

**Comment by paul-tqh-nguyen** [comment]
https://github.com/paul-tqh-nguyen/reuters_topic_labelling/commit/15037a6316880dde313dac416eb69293d87da2a6 implements a gradient descent based oversampling technique. We repeat samples from the training set until the Chi Squared statistic is sufficiently low. 

We'll next investigate how much we have to optimize the Chi Squared statistic (if even at all) to get a better validation F1 score. We might not even see any benefit :/

**Comment by paul-tqh-nguyen** [comment]
https://github.com/paul-tqh-nguyen/reuters_topic_labelling/commit/0e6fae69e4d12c990866b570c1c0df7d879914ac notes that the GPU memory leak problem we've been having has magically disappeared. 

Some initial experimentation has shown that oversampling hasn't improved performance significantly, so we'll close this ticket for now and will re-evaluate when we are writing our documentation.
